site/content/en/docs/developer-guide/plugin_development.md

```diff
@@ -134,11 +134,11 @@ If `WaitAllData` is returned from `DecodeHeaders`, we will:
 
 ![filter manager, with DecodeWholeRequestFilter, buffer the whole request](/images/filtermanager_sub_path.jpg)
 
-Note: `DecodeRequest` is only executed if `DecodeHeaders` returns `WaitAllData`. So if `DecodeRequest` is defined, `DecodeHeaders` must be defined as well.
+Note: `DecodeRequest` is only executed if `DecodeHeaders` returns `WaitAllData`. So if `DecodeRequest` is defined, `DecodeHeaders` must be defined as well. If both `DecodeRequest` and `DecodeData` are defined in the plugin, which method is executed depends on whether `DecodeHeaders` returns `WaitAllData`.
```

Comment:
It's better if we can have a detailed example for this corner case, correct me if I'm wrong in the following example:

Let's say one request body is divided into 3 pieces:
* piece1, Decodeheader return "Continue" , decodeData invoked, decodeRequest skipped.
* piece2, DecodeHeader return "WaitAllData", decodeData skipped, decodeRequest skipped.
* **piece3, Decodeheader return "Continue", decodeData invoked, decodeRequest invoked?**

Reply:
When both `DecodeRequest` and `DecodeData` are defined in the plugin: if `DecodeHeaders` returns `WaitAllData`, only `DecodeRequest` is executed, otherwise, only `DecodeData` is executed.

I will update this to the doc.

Reply:
You didn't answer my question, "decodeHeader" can be invoked several times, so the return result can be different.

Reply:
I get what you mean. No, the DecodeHeader will only run once when the header is received, regardless of how many pieces a request body has. There is only one chance to choose the path.